R/utils.R

```diff
@@ -123,22 +123,48 @@ check_modules_datanames <- function(modules, datanames) {
   recursive_check_datanames <- function(modules, datanames) {
     # check teal_modules against datanames
     if (inherits(modules, "teal_modules")) {
-      sapply(modules$children, function(module) recursive_check_datanames(module, datanames = datanames))
+      result <- lapply(modules$children, function(module) recursive_check_datanames(module, datanames = datanames))
+      result <- result[vapply(result, Negate(is.null), logical(1L))]
+      list(
+        string = do.call(c, as.list(unname(sapply(result, function(x) x$string)))),
+        html = function(with_module_name = TRUE) {
+          tagList(
+            lapply(
+              result,
+              function(x) x$html(with_module_name = with_module_name)
+            )
+          )
+        }
+      )
     } else {
       extra_datanames <- setdiff(modules$datanames, c("all", datanames))
       if (length(extra_datanames)) {
-        sprintf(
-          "Module '%s' uses datanames not available in 'data': (%s) not in (%s)",
-          modules$label,
-          toString(dQuote(extra_datanames, q = FALSE)),
-          toString(dQuote(datanames, q = FALSE))
+        list(
+          string = build_datanames_error_message(
+            modules$label,
+            datanames,
```

Comment:
We enumerate all available datanames in the included object even those irrelevant for the module. If `mod-2` uses only `c("iris", "inexisting")`, then listing `ADSL`, `ADTTE`, etc. seems irrelevant

Reply:
I see it as a helpuful debugging information when the app developer is designing the app, but it will be irrelevant for normal users. So, does it:

1. Makes sense to keep a second part of the message? 
2. Or should we keep the valid datanames and say: "`iris` is available" (for your example)

Reply:
I'm fine with both as far as (1) will be rephrased little bit. `available data for 'mod-2', which has: `. It is not straightforward to what exactly "which has" pointing to, module or data?
I'd prefer: `Dataset(s) ... is/are missing for tab 'mod-2'. Datasets available in data: ...`

Reply:
Ok, here's a new swing with that suggestion, WDYT?

![image](https://github.com/user-attachments/assets/86b9739f-b63e-4e2c-9e04-08e6bec931a7)

Reply:
Added messsage exception when there are no datasets available:

![image](https://github.com/user-attachments/assets/3492e227-310c-437d-b42e-2915827839d5)

<details>

<summary>No datasets available example</summary>

```r
options(
  teal.log_level = "TRACE",
  teal.show_js_log = TRUE,
  # teal.bs_theme = bslib::bs_theme(version = 5),
  shiny.bookmarkStore = "server"
)

pkgload::load_all("teal.data")
pkgload::load_all("teal.slice")
pkgload::load_all("teal")

my_transformers <- list(
  teal_transform_module(
    label = "Add ELO",
    ui = function(id) {
      ns <- NS(id)
      tagList(
        div("Some UI for transform (merge)"),
        actionButton(ns("btn"), "Reload data")
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        eventReactive(input$btn, {
          data() |> within(mtcars <- datasets::mtcars)
        })
      })
    }
  )
)

teal::init(
  data = teal.data::teal_data(iris = iris),
  modules = list(
    example_module("mod-1", datanames = c("mtcars"), transformers = my_transformers),
    example_module("mod-2", datanames = c("iris"))
  )
) |>
  runApp()
```
</details>